Click through migration

dpsutton opened the pull request:
Fixes #15014 

Updates the dashboard card click through behavior.

Would love eyes from @paulrosenzweig and @flamber .

One particular thing to note, is that the original postgres script threw away column info if it didn't have values for `view_as` and `link_template`. 

```sql
FROM (
      SELECT
        key as field,
        value::jsonb->'view_as' as clickType,
        value::jsonb->'link_template' as linkTemplate,
        value::jsonb->'link_text' as linkTextTemplate
      FROM jsonb_each_text(jsonb_merge(card.visualization_settings::jsonb, dashcard.visualization_settings::jsonb)::jsonb->'column_settings')
    ) as json_table
WHERE jsonb_typeof(clickType)!='null' AND jsonb_typeof(linkTemplate)!='null'
```

I've kept this behavior but just wanted to call it out and make sure this was intentional. It would be trivial to preserve these as well.

```clojure
(reduce-kv (fn [m col field-settings]
             (if (and (contains? field-settings "view_as")
                      (contains? field-settings "link_template"))
               (assoc m col
                      {"type"             (field-settings "view_as")
                       "linkType"         "url"
                       "linkTemplate"     (field-settings "link_template")
                       "linkTemplateText" (field-settings "link_text")} )
               m))
           {}
           column-settings)
```

Note this also leaves the old top level stuff on the dashcard after moving it under "click_behavior"
```
"click" "link",
"click_link_template" "http://example.com/{{count}}",
```
The original migration did not remove these. Might help for future that we leave the old datashape? Note we _do_ clobber the `column_settings` so this "original" information is gone though. It would be possible to preserve this just in case in a "column_settings_backup" or other name.

The tests have a good before and after which i'll include here as well:

### Before
```clojure
:card_visualization
{"column_settings"
 {"[\"ref\",[\"field-id\",2]]"
  {"view_as" "link",
   "link_template" "http://example.com/{{ID}}",
   "link_text" "here's an id: {{ID}}"},
  "[\"ref\",[\"field-id\",6]]"
  {"view_as" "link",
   "link_template" "http://example.com//{{id}}",
   "link_text" "here is my id: {{id}}"}},
 "table.pivot_column" "QUANTITY",
 "table.cell_column" "DISCOUNT",
 "click" "link",
 "click_link_template" "http://example.com/{{count}}",
 "graph.dimensions" ["CREATED_AT"],
 "graph.metrics" ["count"],
 "graph.show_values" true},
:dashcard_visualization
{"click" "link",
 "click_link_template"
 "http://localhost:3001/?year={{CREATED_AT}}&cat={{CATEGORY}}&count={{count}}",
 "graph.dimensions" ["CREATED_AT" "CATEGORY"],
 "graph.metrics" ["count"]}
```

### After
```clojure
{"click" "link", ;; note old stuff here
 "click_link_template" ;; old stuff here
 "http://localhost:3001/?year={{CREATED_AT}}&cat={{CATEGORY}}&count={{count}}",
 "graph.dimensions" ["CREATED_AT" "CATEGORY"],
 "graph.metrics" ["count"],
 "click_behavior" ;; new stuff under heading
 {"type" "link",
  "linkType" "url",
  "linkTemplate" "http://example.com/{{count}}"},
 "column_settings" ;; new column settings but these throw away any without the specific info we're looking for
 ;; note none of this keywordizes keys in json parsing since these structures are gross as keywords
 {"[\"ref\",[\"field-id\",2]]"
  {"type" "link",
   "linkType" "url",
   "linkTemplate" "http://example.com/{{ID}}",
   "linkTemplateText" "here's an id: {{ID}}"},
  "[\"ref\",[\"field-id\",6]]"
  {"type" "link",
   "linkType" "url",
   "linkTemplate" "http://example.com//{{id}}",
   "linkTemplateText" "here is my id: {{id}}"}}}
```

dpsutton committed:
Fix click through behavior (#15014)

dpsutton committed:
formatting

dpsutton committed:
Safer get's, author metadata and correct added

dpsutton committed:
Don't return anything if we didn't modify it

dpsutton committed:
Make private and add a docstring

dpsutton commented:
Note: I haven't done this in the UI and then seen that the resulting migration works. I'm gonna back up to 36, do this, and then go back to this branch.

rlotun commented:
also, can we check that if this happens to be applied on the same DB post @flamber's fix that it will work? otherwise if not that's fine - we just have to let the customer know to make sure to apply this on their previous 37 db. thanks for working on it!

camsaul commented on test/metabase/db/data_migrations_test.clj:
```diff
@@ -189,3 +190,85 @@
+            "graph.metrics" ["count"]}}))))
+
+(deftest migrate-click-through-test
+  (let [card-vis     "{\"column_settings\":{\"[\\\"ref\\\",[\\\"field-id\\\",2]]\":{\"view_as\":\"link\",\"link_template\":\"http://example.com/{{ID}}\",\"link_text\":\"here's an id: {{ID}}\"},\"[\\\"ref\\\",[\\\"field-id\\\",6]]\":{\"view_as\":\"link\",\"link_template\":\"http://example.com//{{id}}\",\"link_text\":\"here is my id: {{id}}\"}},\"table.pivot_column\":\"QUANTITY\",\"table.cell_colum ...
```
I'd just write this as a Clojure map and wrap it in `json/generate-string`

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -334,6 +336,64 @@
       (db/update-where! model {:collection_id nil}
         :collection_id (u/get-id new-collection)))))
 
+(defn- fix-click-through
```
A comment referencing the original GitHub issue number would be helpful, so if we don't remember what the point of all this code is, we can easily go back and read the issue/this PR to understand it a bit better

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -334,6 +336,64 @@
+            x
+            ks)))
+
+(defmigration ^{:author "dpsutton" :added "0.38.1"} migrate-click-through
```
It would be helpful to have a comment explaining what's going on above the migration... see the one above it. Right now we all know exactly what it's doing and why, but when new people join the team it might not be so obvious what the purpose of this migration is or even what it's doing. It's better to be able to read a written description then to have to read all the code to try to understand what's going on

camsaul commented on test/metabase/db/data_migrations_test.clj:
```diff
@@ -189,3 +190,85 @@
                  (perms group))))))))
+
+(deftest fix-click-through-test
+  (is (= {:id 1,
```
A `(testing "... (#15014)"` blurb would be super-helpful for these tests. I can't count how many times I've grepped the code-base for `#<issue-number>` to find tests that fixed an issue similar to one I'm working on fixing

camsaul reviewed (approved):
This code looks great to me, the only way it could be made a little better is with a little more documentation so people who join the company in the future can more easily understand the why and what behind this code without having to ask around on Slack

dpsutton commented:
sounds good. i'll document and make sure its idempotent. @rlotun brought up a very good point and want to make sure that's fine. i suspect there might be some issues but i'll verify and test that it works.

dpsutton committed:
Generate json strings rather than long in-line versions

dpsutton committed:
Docstrings and prevent migrating ones that may already be migrated

paulrosenzweig commented on src/metabase/db/data_migrations.clj:
```diff
@@ -334,6 +336,109 @@
+    }
+  column_settings {
+    [ref [field-id 6]] {
+      type:
+      linkType:
+      linkTemplate:
+      linkTemplateText:
```
It's just a comment, but these are also nested in `click_behavior`.

paulrosenzweig commented on src/metabase/db/data_migrations.clj:
```diff
@@ -334,6 +336,109 @@
+                                              (if (and (contains? field-settings "view_as")
+                                                       (contains? field-settings "link_template"))
+                                                (assoc m col
+                                                       {"type"             (get field-settings "view_as")
+                                                        "linkType"         "url"
+                                                        "linkTemplate"     (get field-settings "link_template")
+                                                        "linkTemplateText" (get field-settings "link_text")} )
```
This should be nested under `click_behavior`.

paulrosenzweig reviewed (commented):
> One particular thing to note, is that the original postgres script threw away column info if it didn't have values for view_as and link_template.

We shouldn't do that. Let me look at it a bit more, but we should preserve unrelated settings on both the card and dashcard.

Edit: I looked at it more. I think that case escaped my initial testing of the script. Other column settings on the dashcard should be merged in even if there's no click behavior.

dpsutton committed:
Noisier as it now merges click_behavior on top of existing

dpsutton committed:
update docstring with new nesting

dpsutton committed:
always keep column settings

previously the script was only looking for column settings which
included view_as or link_template. Now we always keep those columns
and merge in the new click_behavior when those keys are present

flamber reviewed (changes requested):
Reference guide:
dashviz = report_dashboardcard.visualization_settings
cardviz = report_card.visualization_settings
CDT = Custom Drill-through (old style, pre-37)
CB = Click Behavior (new style, 37+)

> One particular thing to note, is that the original postgres script threw away column info if it didn't have values for view_as and link_template.

That's how it should work. It should only include the new `"click_behavior":{...}`, since those results are merged together with dashviz.
And it gets the old CDT from cardviz and dashviz, which is also merged (in the right order), since the query would then be much larger if it had to run through each by itself, which we don't care about, since we only want the old CDT - not all the other information, since the frontend automatically reads the visualization_settings and replaces only the added columns.

This PR definitely looks like it's on the right track, but there are some blockers, since we definitely want to make sure that this migration is correct and doesn't overwrite any new CB that might have been put in since 1.37.0

Tested `03b5cf7`
1. It does not include the custom text. It should be `linkTextTemplate` instead of `linkTemplateText`
2. It overwrites any dashviz modifications that are made that changed behavior from cardviz. From what I can tell, it looks like it's only reading cardviz and applying that to dashviz, so any CDT done on the dashboard are not included at all.
3. It includes disabled CDT (`view_as=null`), which then does not work on the new CB and doesn't even fallback to regular Action menu.
4. It copies over the entire cardviz to dashviz (from what I can tell), which means that it will not be possible to make changes (non-CB/CDT) to already custom setting of a question and have that reflected on dashboards until "Reset to defaults" is used on the dashboard card (and that comes with it's own problems using Reset - issue #14919, now that the single-reset was removed #13968). It should only include the newly migrated CB into dashviz.
5. It replaces existing CB, which someone might have now manually setup (and perhaps different from the old CDT) if they upgraded to 1.37.0.
I have updated my SQL to skip anything that already has a new CB: https://gist.github.com/flamber/3212a49ca60cad28ba53272a444d6054

dpsutton committed:
linkTemplateText -> linkTextTemplate

dpsutton committed:
Ensure dashcard goes on top of card viz settings

dpsutton commented:
1. Renamed.
2. swapped merge order so now its `(merge card dashcard)` which will put dashcard on top of card. It is certainly reading dashviz:

```clojure
(db/query {:select    [:dashcard.id
                       [:card.visualization_settings :card_visualization]
                       [:dashcard.visualization_settings :dashcard_visualization]]
           :from      [[:report_dashboardcard :dashcard]]
           :left-join [[:report_card :card] [:= :dashcard.card_id :card.id]]
           :where     [:or
                       [:like
                        :card.visualization_settings "%\"link_template\":%"]
                       [:like
                        :card.visualization_settings "%\"click_link_template\":%"]
                       [:like
                        :dashcard.visualization_settings "%\"link_template\":%"]
                       [:like
                        :dashcard.visualization_settings "%\"click_link_template\":%"]]})
```

3. done
4. done. That means the "dropping columns" issue we thought was a bug was a feature. Now only keep columns if they have a) non-nil value for `"view_as"` and have the key for `"link_template"`
5. this makes sense but it seems a little iffy to make separate checks on the click behavior. Like if there's a top level click_behavior should we still attempt to merge the columns? And then we have the potential to throw out any columns that have the correct shape because that only keeps columns that have the old "bad" shape right? My gut reaction is to just abandon everything if the top level already has a click_behavior. Would love your thoughts on that.

thanks @flamber

dpsutton committed:
Update expected now that dash merges on top of card

dpsutton committed:
drop columns that don't have view_as

so we don't overwrite the top level stuff if there is already a
click_behavior. That signifies that either this migration has run, or
perhaps already has had manual intervention. An open question is if we
extend this to the columns: if there's manual intervention do we
really want to be mucking about in here?

Also, go back to the dropping columns style.

```clojure
(if (and ...)
    (assoc m col (merge field-settings click-behavior))
    m) ;; here we omit the column again in our reduce-kv
```

flamber commented:
Awesome! Taking it for a spin a little later.
As for 5)
There's no time when click_behavior is both at top-level and in column-level. It's at top-level for non-tables, it's at column-level on tables.
I do a check if the click_behavior is present per top-level and column-level individually. If it is, then I skip that top/column-level, so I don't overwrite anything that might have been created since 1.37.0.
Perhaps it's okay just to check if click_behavior is present on dashviz and then skip that in the migration, but then you might not migrate all - example if someone created a table with 8 CDT columns, but on 1.37.0 they manually set 2 columns with CB.

dpsutton commented:
@flamber i think for 5, it should be something like this
- if top level has existing click behavior, do not attempt to compute new click behavior migration for top level
- when iterating through columns:
  - if it has existing click_behavior don't modify that column and just keep it as is
  - if it has old style stuff migration to new click_behavior stuff
  - otherwise let the column fall away

is this correct?

paulrosenzweig commented:
By "let the column fall away" you mean leave the dashcard column settings untouched rather than remove it, right?

flamber commented:
That definitely sounds very close to what I'm doing.

I'm not sure about "otherwise let the column fall away", since that should never happen? Unless you mean when there's an old CDT, which is disabled (view_as!="link"), then I don't count that as an CDT, but simply ignore it.
Sorry, working on two other issues right now, so my brain is a little :exploding_head: of context switching.

paulrosenzweig commented on src/metabase/db/data_migrations.clj:
```diff
@@ -334,6 +336,113 @@
+                                    "linkType"     "url"
+                                    "linkTemplate" (get merged "click_link_template")})
+        has-click-behavior?      (get dashcard "click_behavior")
+        column-settings          (get merged "column_settings")
```
I think this is wrong. If `merge` is shallow, then any dashcard column settings will shadow all card column settings, right?

dpsutton replied:
yes that's correct. is that bad? I have zero domain knowledge and just trying to glean what i can from a sql script. every conversation seems to add more domain knowledge. My guess was that the json merge in the example script was shallow as well.

dpsutton commented:
> By "let the column fall away" you mean leave the dashcard column settings untouched rather than remove it, right?

No. I thought I was told that we ignore all columns that don't have `view_as` on them.

>  > One particular thing to note, is that the original postgres script threw away column info if it didn't have values for view_as and link_template.

> That's how it should work. It should only include the new "click_behavior":{...}, since those results are merged together with dashviz.

paulrosenzweig commented:
Here are some before/after examples. cc @flamber for a sanity check.

# Example 1 - Top level
## Before
### Card
```
{
  some_setting: {foo: 123},
  click_link_template: "http://example.com/{{col_name}}",
  click: "link",
}
```
### Dashcard
```
{
  other_setting: {bar: 123},
}
```
## After
### Card
(unchanged)
### Dashcard
`click: "menu"` turned off the custom drill through, so it's not migrated.
```
{
  other_setting: {bar: 123},
  click_behavior: {
    type: "link",
    linkType: "url",
    linkTemplate: "http://example.com/{{col_name}}",
  }
}
```
---
# Example 2 - Top level disabled in dashcard
## Before
### Card
```
{
  some_setting: {foo: 123},
  click_link_template: "http://example.com/{{col_name}}",
  click: "link",
}
```
### Dashcard
```
{
  other_setting: {bar: 123},
  click_link_template: "http://example.com/{{col_name}}",
  click: "menu",
}
```
## After
### Card
(unchanged)
### Dashcard
`click: "menu"` turned off the custom drill through, so it's not migrated. Dropping `click` and `click_link_template` would be fine but isn't needed.
```
{
  other_setting: {bar: 123},
  click_link_template: "http://example.com/{{col_name}}",
  click: "menu",
}
```
---
# Example 3 - Column settings
## Before
### Card
```
{
  some_setting: {foo: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      view_as: "link",
      link_template: "http://example.com/{{ID}}",
      link_text: "here's an id: {{ID}}"
    },
  },
}
```
### Dashcard
```
{
  other_setting: {bar: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      fun_formatting: "foo",
    },
    "[\"ref\",[\"field-id\",2]]": {
      other_fun_formatting: 123,
    },
  },
}
```
## After
### Card
(unchanged)
### Dashcard
```
{
  other_setting: {bar: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      fun_formatting: "foo",
      click_behavior: {
        type: "link",
        linkType: "url",
        linkTemplate: "http://example.com/{{id}}",
        linkTextTemplate: "here is my id: {{id}}"
      },
    },
    "[\"ref\",[\"field-id\",2]]": {
      other_fun_formatting: 123,
    },
  },
}
```
---
# Example 4 - Manually updated to new behavior
## Before
### Card
```
{
  some_setting: {foo: 123},
  click_link_template: "http://example.com/{{col_name}}",
  click: "link",
}
```
### Dashcard
```
{
  other_setting: {bar: 123},
  click_behavior: {
    type: "link",
    linkType: "url",
    linkTemplate: "http://example.com/{{other_col_name}}",
  }
}
```
## After
### Card
(unchanged)
### Dashcard
```
{
  other_setting: {bar: 123},
  click_behavior: {
    type: "link",
    linkType: "url",
    linkTemplate: "http://example.com/{{other_col_name}}",
  }
}
```

dpsutton commented:
@paulrosenzweig you are Metabae of the month for the month of March

dpsutton committed:
Fundamentally: scour card for click and merge them into dashcard

reshape all click behaviors in both and merge them in.

Still some broken tests at the moment

dpsutton committed:
Remove old style click behavior no longer expected

dpsutton committed:
fixed

dpsutton committed:
Some more comments and remove some silly select keys and when-lets

just compute both keys and then u/select-non-nil-keys and be done with
it

dpsutton committed:
Docstrinc additions about nil

flamber commented:
It's almost there :tada: :trophy: !

1. It seems like if you have manually created new style CB on dashviz, then those are replaced by old style CDT from cardviz.
Mr. Metabae's awesome example 4, which should have the "After: Dashcard" as `(unchanged)` - I have created example 5 below.
2. I would probably leave the old style CDT instead of removing them just in case of traceability and downgrade (forgot backup).
I know we're going to address blobs in the future, where cleanup and versioning will be part of it.

---
# Example 5 - Manually updated to new behavior on Column
## Before
### Card
```
{
  some_setting: {foo: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      view_as: "link",
      link_template: "http://example.com/{{id}}",
      other_special_formatting: "currency",
    },
    "[\"ref\",[\"field-id\",2]]": {
      view_as: "link",
      link_template: "http://example.com/{{something_else}}",
      other_fun_formatting: 0,
    },
  },
}
```
### Dashcard
```
{
  other_setting: {bar: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      click_behavior: {
        type: "link",
        linkType: "url",
        linkTemplate: "http://example.com/{{id}}",
      }
    },
    "[\"ref\",[\"field-id\",2]]": {
      other_fun_formatting: 123,
    },
  },
}
```
## After
### Card
(unchanged)
### Dashcard
(field-id=1 is unchanged, and field-id=2 has the new click_behavior added)
```
{
  other_setting: {bar: 123},
  column_settings: {
    "[\"ref\",[\"field-id\",1]]": {
      click_behavior: {
        type: "link",
        linkType: "url",
        linkTemplate: "http://example.com/{{id}}",
      }
    },
    "[\"ref\",[\"field-id\",2]]": {
      other_fun_formatting: 123,
      click_behavior: {
        type: "link",
        linkType: "url",
        linkTemplate: "http://example.com/{{something_else}}",
      }
    },
  },
}
```

dpsutton committed:
Remove niles, test scenario #5 from flamber

dpsutton committed:
Handle empties and nil _AFTER_ merging

previously had taken care to punt things down to nil, collapse empty
maps, etc. but this is a value that will merge on top of a map rather
than with the map

```clojure
(m/deep-merge {:a {:b :c}} {:a nil})
{:a nil}
;; vs
(m/deep-merge {:a {:b :c}} {:a {}})
{:a {:b :c}}
```

so let the postwalk after the merge clean up empty collections and nil
values and let them blend as normal

dpsutton commented:
Updated! That was basically already working except not removing nils.

flamber commented:
I've tested `a401d19` and it still seems to replace new style CB with the old style CDT. That's example 4 and 5.

I have both Bar and Table that was created with old-CDT on the question with 1.36.
And then I upgraded to 1.37 and manually added new-CB on the dashboard cards.
When I upgrade to `a401d19`, then it replaces the manually added new-CB with the migrated old-CDT.

---

#### question (created on 1.36)

table card:
```
{"column_settings":{"[\"ref\",[\"field-id\",4]]":{"view_as":"link","link_template":"http://localhost/?QCDT&{{CATEGORY}}","link_text":"MyQCDT {{CATEGORY}}","column_title":"QCDT Category"},"[\"ref\",[\"field-id\",6]]":{"view_as":"link","column_title":"QCDT Rating","link_text":"Rating: {{RATING}}","link_template":"http://localhost/?QCDT&{{RATING}}","prefix":"prefix-","suffix":"-suffix"},"[\"ref\",[\"field-id\",5]]":{"view_as":null,"link_text":"QCDT was disabled","link_template":"http://localhost/?QCDT&{{TITLE}}","column_title":"(QCDT disabled) Title"}},"table.pivot_column":"CATEGORY","table.cell_column":"PRICE"}
```

bar card:
```
{"click":"link","click_link_template":"http://localhost/?QCDT&{{CATEGORY}}","graph.dimensions":["CATEGORY"],"graph.metrics":["count"]}
```

---

#### 1.37 dashcard

table:
```
{"table.cell_column":"PRICE","table.pivot_column":"CATEGORY","column_settings":{"[\"ref\",[\"field-id\",5]]":{"view_as":null,"link_text":"QCDT was disabled","link_template":"http://localhost/?QCDT&{{TITLE}}","column_title":"(QCDT disabled) Title"},"[\"ref\",[\"field-id\",4]]":{"view_as":"link","link_template":"http://localhost/?QCDT&{{CATEGORY}}","link_text":"MyQCDT {{CATEGORY}}","column_title":"QCDT Category","click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?CB&{{CATEGORY}}","linkTextTemplate":"MyCB {{CATEGORY}}"}},"[\"ref\",[\"field-id\",6]]":{"view_as":"link","column_title":"QCDT Rating","link_text":"Rating: {{RATING}}","link_template":"http://localhost/?QCDT&{{RATING}}","prefix":"prefix-","suffix":"-suffix"}},"card.title":"Table with QCDT - MANUALLY ADDED CB 37"}
```

bar:
```
{"click":"link","click_link_template":"http://localhost/?QCDT&{{CATEGORY}}","graph.dimensions":["CATEGORY"],"graph.metrics":["count"],"card.title":"Bar with QCDT - MANUALLY ADDED CB 37","click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?CB&{{CATEGORY}}"}}
```

---

#### a401d19 dashcard

table:
```
{"column_settings":{"[\"ref\",[\"field-id\",4]]":{"click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?QCDT&{{CATEGORY}}","linkTextTemplate":"MyQCDT {{CATEGORY}}"},"column_title":"QCDT Category"},"[\"ref\",[\"field-id\",6]]":{"click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?QCDT&{{RATING}}","linkTextTemplate":"Rating: {{RATING}}"},"column_title":"QCDT Rating","prefix":"prefix-","suffix":"-suffix"},"[\"ref\",[\"field-id\",5]]":{"link_text":"QCDT was disabled","link_template":"http://localhost/?QCDT&{{TITLE}}","column_title":"(QCDT disabled) Title"}},"table.cell_column":"PRICE","table.pivot_column":"CATEGORY","card.title":"Table with QCDT - MANUALLY ADDED CB 37"}
```

bar:
```
{"graph.dimensions":["CATEGORY"],"graph.metrics":["count"],"card.title":"Bar with QCDT - MANUALLY ADDED CB 37","click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?QCDT&{{CATEGORY}}"}}
```

---

#### my script dashcard

table:
```
{"card.title": "Table with QCDT - MANUALLY ADDED CB 37", "column_settings": {"[\"ref\",[\"field-id\",4]]": {"view_as": "link", "link_text": "MyQCDT {{CATEGORY}}", "column_title": "QCDT Category", "link_template": "http://localhost/?QCDT&{{CATEGORY}}", "click_behavior": {"type": "link", "linkType": "url", "linkTemplate": "http://localhost/?CB&{{CATEGORY}}", "linkTextTemplate": "MyCB {{CATEGORY}}"}}, "[\"ref\",[\"field-id\",5]]": {"link_text": "QCDT was disabled", "column_title": "(QCDT disabled) Title", "link_template": "http://localhost/?QCDT&{{TITLE}}"}, "[\"ref\",[\"field-id\",6]]": {"prefix": "prefix-", "suffix": "-suffix", "view_as": "link", "link_text": "Rating: {{RATING}}", "column_title": "QCDT Rating", "link_template": "http://localhost/?QCDT&{{RATING}}", "click_behavior": {"type": "link", "linkType": "url", "linkTemplate": "http://localhost/?QCDT&{{RATING}}", "linkTextTemplate": "Rating: {{RATING}}"}}}, "table.cell_column": "PRICE", "table.pivot_column": "CATEGORY"}
```

bar:
```
{"click":"link","click_link_template":"http://localhost/?QCDT&{{CATEGORY}}","graph.dimensions":["CATEGORY"],"graph.metrics":["count"],"card.title":"Bar with QCDT - MANUALLY ADDED CB 37","click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?CB&{{CATEGORY}}"}}
```

dpsutton commented:
@flamber i don't know what table and bar refer to here. My mental model is solely dashcard and card. Can you tell me which serves as which in this example?

flamber commented:
That's just visualization types - Table and Bar - Table uses column-level, and Bar uses top-level.
Everything except "question (created on 1.36)" is dashcard

dpsutton commented:
> That's just visualization types - Table and Bar - Table uses column-level, and Bar uses top-level.
> Everything except "question (created on 1.36)" is dashcard

oh i see. these are all dashcards, with no card visualization settings to supplement them? So just checking the fixes on the dashcard?

flamber commented:
I have updated the text - @paulrosenzweig can you help explaining - I've been working 10 hours already, so I cannot do any better

dpsutton committed:
Deep merge and prioritize new shape in card and dashcard

```clojure
(m/deep-merge fixed-card
              fixed-dashcard
              (existing-fixed card)
              (existing-fixed dashcard))
```

This is the change here essentially. Fix the card and dashcard, look
for existing new shapes on the card and dashcard as well

dpsutton commented:
So i think i pieced this together. There was no card information provided but i think we need to allow existing new shapes click behavior on cards to land on top of old style click behavior that we migrate:

```clojure
(let [dash {"column_settings"
            {"[\"ref\",[\"field-id\",4]]"
             {"view_as"       "link"
              "link_template" "http://from/dash" ;; old style click behavior on dashcard
              "link_text"     "from dash"
              "column_title"  "column title from dash"}}}
      card {"column_settings"
            {"[\"ref\",[\"field-id\",4]]"
             {"click_behavior" ;; new style click behavior on card
              {"type"             "link",
               "linkType"         "url",
               "linkTemplate"     "http://from/card",
               "linkTextTemplate" "from card"}
              "column_title" "column title from card"}}}]
  (is (= {"column_settings"
          {"[\"ref\",[\"field-id\",4]]"
           {"click_behavior"
            {"type" "link",
             "linkType" "url",
             "linkTemplate" "http://from/card", ;; click behavior survives from the card since it was new style
             "linkTextTemplate" "from card"},
            "column_title" "column title from dash"}}} ;; other information on the dashcard survives
         (migrate card dash))))
```

Here the card has a new shape under `click_behavior` so that should win over the old style stuff on the dashcard. However, things like `column_title` from the dash should stay around as well.

@flamber can you confirm that this understanding is correct? And that this new behavior matches your expectations? The logic in code is quite succinct, with things on the right merging on top of things on the left:

```clojure
(m/deep-merge fixed-card fixed-dashcard (existing-fixed card) (existing-fixed dashcard))
```

flamber commented:
I don't understand Clojure, but `dashcard` overrules `card` - that's how you're allowed to change the title of a question on the dashboard.
It's not possible to have the new `click_behavior` on `card` - it's a `dashcard`-only feature - whereas the old style could exists on both `dashcard` and `card`, with `dashcard` overruling `card`'s behavior.

Slightly stolen from the 3 rules of AI:
1. Never modify `card`
2. Only update `dashcard`, when there is not an `click_behavior` already for viz's root-level or table's column-level (each). In short; never overwrite `click_behavior` - and never contradict rule 1.
3. Get the old behavior from `card` and `dashcard`, where `dashcard` will be the ruler, when old behavior was valid (ie `view_as = "link"`), and generate new `click_behavior` on `dashcard` - and never contradict rule 1 and 2.

dpsutton commented:
Then I don't understand your examples here. You start with this as Step 1:

Here's your first example for "table card":
```clojure
{"column_settings"
 {"[\"ref\",[\"field-id\",4]]"
  {"view_as" "link"
   "link_template" "http//localhost/?QCDT&{{CATEGORY}}"
   "link_text" "MyQCDT {{CATEGORY}}"
   "column_title" "QCDT Category"}
  "[\"ref\",[\"field-id\",6]]"
  {"view_as" "link"
   "column_title" "QCDT Rating"
   "link_text" "Rating {{RATING}}"
   "link_template" "http//localhost/?QCDT&{{RATING}}"
   "prefix" "prefix-"
   "suffix" "-suffix"}
  "[\"ref\",[\"field-id\",5]]"
  {"view_as" nil
   "link_text" "QCDT was disabled"
   "link_template" "http//localhost/?QCDT&{{TITLE}}"
   "column_title" "(QCDT disabled) Title"}}
 "table.pivot_column" "CATEGORY"
 "table.cell_column" "PRICE"}
```

The code currently migrates this to:
```clojure
{"column_settings"
 {"[\"ref\",[\"field-id\",4]]"
  {"column_title" "QCDT Category",
   "click_behavior"
   {"type" "link",
    "linkType" "url",
    "linkTemplate" "http//localhost/?QCDT&{{CATEGORY}}",
    "linkTextTemplate" "MyQCDT {{CATEGORY}}"}},
  "[\"ref\",[\"field-id\",6]]"
  {"column_title" "QCDT Rating",
   "prefix" "prefix-",
   "suffix" "-suffix",
   "click_behavior"
   {"type" "link",
    "linkType" "url",
    "linkTemplate" "http//localhost/?QCDT&{{RATING}}",
    "linkTextTemplate" "Rating {{RATING}}"}},
  "[\"ref\",[\"field-id\",5]]"
  {"link_text" "QCDT was disabled",
   "link_template" "http//localhost/?QCDT&{{TITLE}}",
   "column_title" "(QCDT disabled) Title"}},
 "table.pivot_column" "CATEGORY",
 "table.cell_column" "PRICE"}
```

Can you tell me what is wrong here? I don't see how you came up with 

```clojure
{"card.title" "Table with QCDT - MANUALLY ADDED CB 37"
 "column_settings"
 {"[\"ref\",[\"field-id\",4]]"
  {"view_as" "link"
   "link_text" "MyQCDT {{CATEGORY}}"
   "column_title" "QCDT Category"
   "link_template" "http//localhost/?QCDT&{{CATEGORY}}"
   "click_behavior"
   {"type" "link"
    "linkType" "url"
    "linkTemplate" "http//localhost/?CB&{{CATEGORY}}"
    "linkTextTemplate" "MyCB {{CATEGORY}}"}}
  "[\"ref\",[\"field-id\",5]]"
  {"link_text" "QCDT was disabled"
   "column_title" "(QCDT disabled) Title"
   "link_template" "http//localhost/?QCDT&{{TITLE}}"}
  "[\"ref\",[\"field-id\",6]]"
  {"prefix" "prefix-"
   "suffix" "-suffix"
   "view_as" "link"
   "link_text" "Rating {{RATING}}"
   "column_title" "QCDT Rating"
   "link_template" "http//localhost/?QCDT&{{RATING}}"
   "click_behavior"
   {"type" "link"
    "linkType" "url"
    "linkTemplate" "http//localhost/?QCDT&{{RATING}}"
    "linkTextTemplate" "Rating {{RATING}}"}}}
 "table.cell_column" "PRICE"
 "table.pivot_column" "CATEGORY"}
```

diffing these two structures, yours first and the result of migrating second, yields:

```clojure
[;; things that are in yours and not mine
 {"column_settings"
  {"[\"ref\",[\"field-id\",6]]"
   {"view_as" "link",                   ;; you leave in the original stuff i remove it
    "link_text" "Rating {{RATING}}",
    "link_template" "http//localhost/?QCDT&{{RATING}}"},
   "[\"ref\",[\"field-id\",4]]"
   {"click_behavior"
    {"linkTextTemplate" "MyCB {{CATEGORY}}",
     "linkTemplate" "http//localhost/?CB&{{CATEGORY}}"},  ;; these two link infos seem not connected
    "view_as" "link",
    "link_text" "MyQCDT {{CATEGORY}}",
    "link_template" "http//localhost/?QCDT&{{CATEGORY}}"}},
  "card.title" "Table with QCDT - MANUALLY ADDED CB 37"} ;; this card title came from now where
 ;; things that are in mine and not yours
 {"column_settings"
  {"[\"ref\",[\"field-id\",4]]"
   {"click_behavior"
    {"linkTextTemplate" "MyQCDT {{CATEGORY}}", ;; i have link stuff from the original you don't
     "linkTemplate" "http//localhost/?QCDT&{{CATEGORY}}"}}}}]
```

@flamber  Can you please help me understand what this data migration should be doing that it is currently not doing?

As far as I can tell the only difference between these is that you are leaving in the `"view_as"`, `"link_text"`, and `"link_template"` and you have a completely different click behavior for field-id 4. Is it important to leave in the old style stuff ?

flamber commented:
Perhaps @paulrosenzweig can assist with explaining this in a different way ?

**Before migration:**

Create a question (card) on 1.36:
```
{"column_settings":{"[\"ref\",[\"field-id\",4]]":{"view_as":"link","link_template":"http://localhost/?QCDT&{{CATEGORY}}","link_text":"MyQCDT {{CATEGORY}}","column_title":"QCDT Category"},"[\"ref\",[\"field-id\",6]]":{"view_as":"link","column_title":"QCDT Rating","link_text":"Rating: {{RATING}}","link_template":"http://localhost/?QCDT&{{RATING}}","prefix":"prefix-","suffix":"-suffix"},"[\"ref\",[\"field-id\",5]]":{"view_as":null,"link_text":"QCDT was disabled","link_template":"http://localhost/?QCDT&{{TITLE}}","column_title":"(QCDT disabled) Title"}},"table.pivot_column":"CATEGORY","table.cell_column":"PRICE"}
```
Add it to a dashboard (dashcard) and manually add new-style `click_behavior` to one of the columns on 1.37:
```
{"table.cell_column":"PRICE","table.pivot_column":"CATEGORY","column_settings":{"[\"ref\",[\"field-id\",5]]":{"view_as":null,"link_text":"QCDT was disabled","link_template":"http://localhost/?QCDT&{{TITLE}}","column_title":"(QCDT disabled) Title"},"[\"ref\",[\"field-id\",4]]":{"view_as":"link","link_template":"http://localhost/?QCDT&{{CATEGORY}}","link_text":"MyQCDT {{CATEGORY}}","column_title":"QCDT Category","click_behavior":{"type":"link","linkType":"url","linkTemplate":"http://localhost/?CB&{{CATEGORY}}","linkTextTemplate":"MyCB {{CATEGORY}}"}},"[\"ref\",[\"field-id\",6]]":{"view_as":"link","column_title":"QCDT Rating","link_text":"Rating: {{RATING}}","link_template":"http://localhost/?QCDT&{{RATING}}","prefix":"prefix-","suffix":"-suffix"}},"card.title":"Table with QCDT - MANUALLY ADDED CB 37"}
```

**After migration:**

This is what the dashcard should look like:
It should **not** replace the manually added `click_behavior` from 1.37
It should add valid `click_behavior` for other columns (or top-level if non-table)
```
{"card.title": "Table with QCDT - MANUALLY ADDED CB 37", "column_settings": {"[\"ref\",[\"field-id\",4]]": {"view_as": "link", "link_text": "MyQCDT {{CATEGORY}}", "column_title": "QCDT Category", "link_template": "http://localhost/?QCDT&{{CATEGORY}}", "click_behavior": {"type": "link", "linkType": "url", "linkTemplate": "http://localhost/?CB&{{CATEGORY}}", "linkTextTemplate": "MyCB {{CATEGORY}}"}}, "[\"ref\",[\"field-id\",5]]": {"link_text": "QCDT was disabled", "column_title": "(QCDT disabled) Title", "link_template": "http://localhost/?QCDT&{{TITLE}}"}, "[\"ref\",[\"field-id\",6]]": {"prefix": "prefix-", "suffix": "-suffix", "view_as": "link", "link_text": "Rating: {{RATING}}", "column_title": "QCDT Rating", "link_template": "http://localhost/?QCDT&{{RATING}}", "click_behavior": {"type": "link", "linkType": "url", "linkTemplate": "http://localhost/?QCDT&{{RATING}}", "linkTextTemplate": "Rating: {{RATING}}"}}}, "table.cell_column": "PRICE", "table.pivot_column": "CATEGORY"}
```

dpsutton commented:
@flamber my tests pass on this case. The only difference is that you leave `"view_as"`, `"link_text"`, and `"link_template"` in the column settings map.

```clojure
(testing "flamber case"
  (let [card {"column_settings"
              {"[\"ref\",[\"field-id\",4]]"
               {"view_as" "link"
                "link_template" "http//localhost/?QCDT&{{CATEGORY}}"
                "link_text" "MyQCDT {{CATEGORY}}"
                "column_title" "QCDT Category"}
               "[\"ref\",[\"field-id\",6]]"
               {"view_as" "link"
                "column_title" "QCDT Rating"
                "link_text" "Rating {{RATING}}"
                "link_template" "http//localhost/?QCDT&{{RATING}}"
                "prefix" "prefix-"
                "suffix" "-suffix"}
               "[\"ref\",[\"field-id\",5]]"
               {"view_as" nil
                "link_text" "QCDT was disabled"
                "link_template" "http//localhost/?QCDT&{{TITLE}}"
                "column_title" "(QCDT disabled) Title"}}
              "table.pivot_column" "CATEGORY"
              "table.cell_column""PRICE"}
        dash {"table.cell_column" "PRICE"
              "table.pivot_column" "CATEGORY"
              "column_settings"
              {"[\"ref\",[\"field-id\",5]]"
               {"view_as" nil
                "link_text" "QCDT was disabled"
                "link_template" "http//localhost/?QCDT&{{TITLE}}"
                "column_title" "(QCDT disabled) Title"}
               "[\"ref\",[\"field-id\",4]]"
               {"view_as" "link"
                "link_template" "http//localhost/?QCDT&{{CATEGORY}}"
                "link_text" "MyQCDT {{CATEGORY}}"
                "column_title" "QCDT Category"
                "click_behavior"
                {"type" "link"
                 "linkType" "url"
                 "linkTemplate" "http//localhost/?CB&{{CATEGORY}}"
                 "linkTextTemplate" "MyCB {{CATEGORY}}"}}
               "[\"ref\",[\"field-id\",6]]"
               {"view_as" "link"
                "column_title" "QCDT Rating"
                "link_text" "Rating {{RATING}}"
                "link_template" "http//localhost/?QCDT&{{RATING}}"
                "prefix" "prefix-"
                "suffix" "-suffix"}}
              "card.title" "Table with QCDT - MANUALLY ADDED CB 37"}]
    (is (= {"card.title" "Table with QCDT - MANUALLY ADDED CB 37"
            "column_settings"
            {"[\"ref\",[\"field-id\",4]]"
             {"column_title" "QCDT Category"
              "click_behavior"
              {"type" "link"
               "linkType" "url"
               "linkTemplate" "http//localhost/?CB&{{CATEGORY}}"
               "linkTextTemplate" "MyCB {{CATEGORY}}"}}
             "[\"ref\",[\"field-id\",5]]"
             {"link_text" "QCDT was disabled"
              "column_title" "(QCDT disabled) Title"
              "link_template" "http//localhost/?QCDT&{{TITLE}}"}
             "[\"ref\",[\"field-id\",6]]"
             {"prefix" "prefix-"
              "suffix" "-suffix"
              "column_title" "QCDT Rating"
              "click_behavior"
              {"type" "link"
               "linkType" "url"
               "linkTemplate" "http//localhost/?QCDT&{{RATING}}"
               "linkTextTemplate" "Rating {{RATING}}"}}}
            "table.cell_column" "PRICE"
            "table.pivot_column" "CATEGORY"}
           (migrate card dash)))))
```

So can you confirm things are good? Or can you confirm that these keys you are leaving in are necessary to be left in?

I think you were originally getting at this peculiarity:
```clojure
{"[\"ref\",[\"field-id\",4]]"
 {"view_as" "link"
  "link_template" "http//localhost/?QCDT&{{CATEGORY}}"
  "link_text" "MyQCDT {{CATEGORY}}"
  "column_title" "QCDT Category"
  "click_behavior"
  {"type" "link"
   "linkType" "url"
   "linkTemplate" "http//localhost/?CB&{{CATEGORY}}"
   "linkTextTemplate" "MyCB {{CATEGORY}}"}}}
```

here we have a click behavior and an old style. And its important that the fixed click behavior survives on top of the migrated click behavior on field-id 4. Is that correct?

dpsutton commented:
a much more concise example of this behavior:

```clojure
(testing "If there is migration eligible on dash but also new style on dash, new style wins"
  (let [dash {"column_settings"
              {"[\"ref\",[\"field-id\",4]]"
               {"view_as"       "link"
                "link_template" "http://old" ;; this stuff could be migrated
                "link_text"     "old"
                "column_title"  "column title"
                "click_behavior"
                {"type" "link",
                 "linkType" "url", ;; but there is already a new style and it wins
                 "linkTemplate" "http://new",
                 "linkTextTemplate" "new"}}}}]
    (is (= {"column_settings"
            {"[\"ref\",[\"field-id\",4]]"
             {"click_behavior"
              {"type" "link",
               "linkType" "url",
               "linkTemplate" "http://new",
               "linkTextTemplate" "new"},
              "column_title" "column title"}}}
           (migrate nil dash)))))
```

dpsutton committed:
Don't merge existing fix from card. can't happen

this change though did help out. we want to resolve when there are old
style and new style on the same dashcard. In this case, the new style
should persist:

```clojure
(testing "If there is migration eligible on dash but also new style on dash, new style wins"
  (let [dash {"column_settings"
              {"[\"ref\",[\"field-id\",4]]"
               {"view_as"       "link"
                "link_template" "http://old" ;; this stuff could be migrated
                "link_text"     "old"
                "column_title"  "column title"
                "click_behavior"
                {"type" "link",
                 "linkType" "url", ;; but there is already a new style and it wins
                 "linkTemplate" "http://new",
                 "linkTextTemplate" "new"}}}}]
    (is (= {"column_settings"
            {"[\"ref\",[\"field-id\",4]]"
             {"click_behavior"
              {"type" "link",
               "linkType" "url",
               "linkTemplate" "http://new",
               "linkTextTemplate" "new"},
              "column_title" "column title"}}}
           (migrate nil dash)))))
```

note the column settings for field-id 4 have the old style and the new
style but the result is that the new style survives. This happens from
the `(m/deep-merge fixed-card fixed-dashcard (existing-fixed
dashcard))` as the existing is put on top

paulrosenzweig commented:
> Or can you confirm that these keys you are leaving in are necessary to be left in?

Leaving in `view_as`, `link_template`, etc is not necessary to correctly migrate the data. It will work fine with or without those keys. @flamber pointed out that's probably best to leave around in case anything goes wrong but that's a forensics rather than functionality need.

dpsutton committed:
Don't strip original values

dpsutton committed:
propagate old info in migration test

flamber reviewed (approved):
Tested `5ec60bf` and it does exactly what it's supposed to :tada: :+1: 
Not replacing existing `click_behavior` (and also keeps old-style just for traceability)
